File: Strings/921_min_add_to_make_valid.cpp
#include <iostream>
#include <string>


// O(N)
class Solution{
public:
    int minAddToMakeValid(const std::string& S) {
        unsigned open = 0, closed = 0;
        
        for (char cur_p : S){
            if (cur_p == '(') open++;
            else if (cur_p == ')' && open > 0) open--;
            else closed++;
        }
        return open + closed;
    }
};


int main(){
    std::string s;
    std::cout << "String with characters '(', ')', '{', '}': ";
    std::cin >> s;

    std::cout << std::boolalpha;
    std::cout << "Minimum add to make parentheses valid: " << Solution().minAddToMakeValid(s) << std::endl;
    return 0;
}

File: Strings/921_min_add_to_make_valid.py
# O(N)
class Solution(object):
    def minAddToMakeValid(self, S):
        opened, closed = 0, 0
        
        for cur_p in S:
            if cur_p == '(': 
                opened += 1
            elif cur_p == ')' and opened > 0:
                opened -= 1
            else:
                closed += 1
        
        return opened + closed


if __name__ == '__main__':
    s = input("String with characters '(', ')', '{', '}': ")

    print('Minimum add to make parentheses valid: ', Solution().minAddToMakeValid(s))
